### PR TITLE
Convert AxHost IPersistX Interface to Cswin32

### PR DESCRIPTION
**Pull request by lonitra**
- Converted rest of AxHost interfaces (IPersistX) to cswin32 version, which prompted for the conversion of other relating interfaces: `IStream`,`IStorage`,`IPropertyBag`,`IPicture`,`IPictureDisp`
- Added helpers to ComHelpers.cs to get IDispatch property and get scoped pointers
- Convert IStream vtable to the new pattern
- Remove PictureWrapper
- Create real mocks in Com2PictureConverterTests since IPicture can no longer be mocked using moq

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/8052)

### CODE REVIEW

File: src/System.Windows.Forms.Primitives/src/Interop/Ole32/Interop.GPStream.cs [hunk 1]
```diff
@@ -153,69 +172,76 @@ public unsafe void Seek(long dlibMove, SeekOrigin dwOrigin, ulong* plibNewPositi
...
                     // Caller wants a name
-                    pstatstg.AllocName(_dataStream is FileStream fs ? fs.Name : _dataStream.ToString());
+                    pstatstg->pwcsName = (char*)Marshal.StringToCoTaskMemUni(_dataStream is FileStream fs ? fs.Name : _dataStream.ToString());
                 }
+
+                return HRESULT.S_OK;
             }
 
-            public unsafe void Write(byte* pv, uint cb, uint* pcbWritten)
+            /// Returns <see cref="HRESULT.STG_E_INVALIDFUNCTION"/> as a documented way to say we don't support locking
```
Comment (JeremyKuhne) [inline src/System.Windows.Forms.Primitives/src/Interop/Ole32/Interop.GPStream.cs:216]: nit: I'd just do a regular comment here, cref'ing STG_E_INVALIDFUNCTION doesn't add much. I'm also not clear on what the rules are for XML comments- I would have thought that having an enclosing tag would have been required?

File: src/System.Windows.Forms.Primitives/src/Windows/Win32/Foundation/ComHelpers.cs [hunk 2]
```diff
@@ -33,5 +34,18 @@ internal static unsafe class ComHelpers
                 ((IUnknown*)release)->Release();
             }
         }
+
+        public static HRESULT InvokePictureDisp(
```
Comment (JeremyKuhne) [inline src/System.Windows.Forms.Primitives/src/Windows/Win32/Foundation/ComHelpers.cs:38]: This should be generic and take `IDispatch*`. I'd do something like this:

`public static HRESULT GetDispatchProperty(IDispatch* dispatch, uint dispid, out VARIANT result)`

**Review by JeremyKuhne — CHANGES_REQUESTED** [review]
Added some comments. I'll connect with you to see if we can remove the PictureWrapper.

File: src/System.Windows.Forms.Primitives/src/Interop/Ole32/Interop.GPStream.cs [hunk 3]
```diff
@@ -85,33 +104,36 @@ public unsafe void CopyTo(IStream pstm, ulong cb, ulong* pcbRead, ulong* pcbWrit
...
 
-            public unsafe void Seek(long dlibMove, SeekOrigin dwOrigin, ulong* plibNewPosition)
-            {
-                long position = _virtualPosition;
-                if (_virtualPosition == -1)
-                {
-                    position = _dataStream.Position;
-                }
+            HRESULT Com.IStream.Interface.Read(void* pv, uint cb, uint* pcbRead) =>
+                ((Com.ISequentialStream.Interface)this).Read(pv, cb, pcbRead);
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/src/Interop/Ole32/Interop.GPStream.cs:130]: It's customary to have `=>` on the new line
```suggestion
            HRESULT Com.IStream.Interface.Read(void* pv, uint cb, uint* pcbRead)
                => ((Com.ISequentialStream.Interface)this).Read(pv, cb, pcbRead);
```

File: src/System.Windows.Forms.Primitives/src/Interop/Ole32/Interop.GPStream.cs [hunk 4]
```diff
@@ -153,69 +175,76 @@ public unsafe void Seek(long dlibMove, SeekOrigin dwOrigin, ulong* plibNewPositi
...
             }
 
-            public HRESULT UnlockRegion(ulong libOffset, ulong cb, uint dwLockType)
-            {
-                // Documented way to say we don't support locking
-                return HRESULT.STG_E_INVALIDFUNCTION;
-            }
-
-            public Stream GetDataStream() => _dataStream;
+            HRESULT Com.IStream.Interface.Write(void* pv, uint cb, uint* pcbWritten) =>
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/src/Interop/Ole32/Interop.GPStream.cs:246]: ditto

**Review by JeremyKuhne — APPROVED** [review]
🔥

**Review by JeremyKuhne — APPROVED** [review]

**Pull request merged**